src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_NullCoalescingAssignmentOperator.cs

```diff
@@ -14,35 +16,143 @@ public override BoundNode VisitNullCoalescingAssignmentOperator(BoundNullCoalesc
+                                          SpecialMember.System_Nullable_T_GetValueOrDefault,
+                                          out var getValueOrDefault))
+                {
+                    return BadExpression(node);
+                }
+
+                if (!TryGetNullableMethod(leftOperand.Syntax,
+                                          leftOperand.Type,
+                                          SpecialMember.System_Nullable_T_get_HasValue,
+                                          out var hasValue))
+                {
+                    return BadExpression(node);
+                }
+
+                // If null coalescing assignment is supported in expression trees, the below code
+                // will need to be updated to support property accesses as well as calls. Currently,
+                // MakeRValue will never return a BoundPropertyAccess except in expression trees.
+                Debug.Assert(!_inExpressionLambda && lhsRead.Kind != BoundKind.PropertyAccess);
+
+                // If lhsRead is a call, such as to a property accessor, save the result off to a temp. This doesn't affect
+                // the standard ??= case because it only uses lhsRead once.
+                if (lhsRead.Kind == BoundKind.Call)
+                {
+                    var lhsTemp = _factory.StoreToTemp(lhsRead, out var store, kind: SynthesizedLocalKind.Spill);
+                    stores.Add(store);
+                    temps.Add(lhsTemp.LocalSymbol);
+                    lhsRead = lhsTemp;
+                }
+
+                // tmp = lhsRead.GetValueOrDefault();
+                var tmp = _factory.StoreToTemp(BoundCall.Synthesized(leftOperand.Syntax, lhsRead, getValueOrDefault),
+                                               out var getValueOrDefaultStore,
+                                               kind: SynthesizedLocalKind.Spill);
+
+                temps.Add(tmp.LocalSymbol);
+
+                // tmp = loweredRight;
+                var tmpAssignment =
+                    MakeAssignmentOperator(node.Syntax, tmp, loweredRight, node.Type, used: true, isChecked: false, isCompoundAssignment: false);
+
+                // transformedLhs = tmp;
+                var transformedLhsAssignment =
+                    MakeAssignmentOperator(
+                        node.Syntax,
+                        transformedLHS,
+                        MakeConversionNode(tmp, transformedLHS.Type, @checked: false),
+                        node.LeftOperand.Type,
+                        used: true,
+                        isChecked: false,
+                        isCompoundAssignment: false); ;
+
+                // !lhsRead.HasValue
+                var lhsReadHasValue = BoundCall.Synthesized(leftOperand.Syntax, lhsRead, hasValue);
+                var lhsReadHasValueNegation =
+                    MakeUnaryOperator(UnaryOperatorKind.BoolLogicalNegation, leftOperand.Syntax, method: null, lhsReadHasValue, lhsReadHasValue.Type);
```

Comment:
>MakeUnaryOperator [](start = 20, length = 17)

Can this be `_factory.Not(...)`?

Reply:
I didn't know about these, thanks.

---
In reply to: [293096260](https://github.com/dotnet/roslyn/pull/36329#discussion_r293096260) [](ancestors = 293096260)